File: java/beer-song/src/main/java/BeerSong.java
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class BeerSong {
  String verse(int verse) {
    if (verse == 0) {
      return """
				No more bottles of beer on the wall, no more bottles of beer.
				Go to the store and buy some more, 99 bottles of beer on the wall.

				""";
    }
    if (verse == 1) {
      return """
				1 bottle of beer on the wall, 1 bottle of beer.
				Take it down and pass it around, no more bottles of beer on the wall.

				""";
    }

    return String.format(
        """
				%d bottles of beer on the wall, %d bottles of beer.
				Take one down and pass it around, %d bottle%s of beer on the wall.

				""",
        verse, verse, verse - 1, (verse == 2) ? "" : "s");
  }

  String sing(int startVerse, int verseNum) {
    return IntStream.range(0, verseNum)
        .map(i -> startVerse - i)
        .mapToObj(this::verse)
        .collect(Collectors.joining(""));
  }

  String singSong() {
    return sing(99, 100);
  }
}

File: cpp/beer-song/beer_song.cpp
#include "beer_song.h"

namespace beer_song
{
std::string verse(int n)
{
    if (n == 0)
    {
        return "No more bottles of beer on the wall, no more bottles of beer.\n"
               "Go to the store and buy some more, 99 bottles of beer on the wall.\n";
    }
    else if (n == 1)
    {
        return "1 bottle of beer on the wall, 1 bottle of beer.\n"
               "Take it down and pass it around, no more bottles of beer on the wall.\n";
    }

    return std::to_string(n) + " bottles of beer on the wall, " + std::to_string(n) + " bottles of beer.\n" +
           "Take one down and pass it around, " + std::to_string(n - 1) + " bottle" + ((n == 2) ? "" : "s") + " of beer on the wall.\n";
}

std::string sing(int n1, int n2)
{
    std::string result;
    for (int i{n1}; i >= n2; --i)
    {
        if (!result.empty())
        {
            result += '\n';
        }

        result += verse(i);
    }

    return result;
}
} // namespace beer_song
